ucp/_libs/ucx_api.pyx

```diff
@@ -102,21 +103,77 @@ def get_ucx_version():
     return (a, b, c)
 
 
-cdef class UCXContext:
-    """Python representation of `ucp_context_h`"""
+def _handle_finalizer_wrapper(children, handle_finalizer, handle_as_int, *extra_args):
+    for weakref_to_child in children:
+        child = weakref_to_child()
+        if child is not None:
+            child.close()
+    handle_finalizer(handle_as_int, *extra_args)
+
+
+cdef class UCXObject:
+    """Base class for UCX classes
+
+    This bass class streamline the cleanup of UCX objects and reduce duplicate code.
+    """
     cdef:
         object __weakref__
+        object _finalizer
+        list _children
+
+    def __cinit__(self):
+        # The finalizer that can be called multiple times but only
+        # evoke the finalizer funciont once.
+        # Is None when the underlying UCX handle hasen't been initialized.
+        self._finalizer = None
+        # List of weak references of UCX objects that make use of this object
+        self._children = []
+
+    def close(self):
+        """Close the object and free the underlying UCX handle.
+        Does nothing if the object is already closed
+        """
+        if self.initialized:
+            self._finalizer()
```

Comment:
Maybe `_finalizer` should just be expected to be idempotent. That way if we call it more than once nothing happens. WDYT?

Reply:
I did that initially but it requires a dummy finalizer class for the case when `add_handle_finalizer()` hasn't been called yet. 
I concluded that this solution was less code and more readable.